### PR TITLE
feat: escrow transaction builder

### PR DESCRIPTION
**Pull request by KolevDarko**
The transaction builder now generates an approval transaction for the escrow contract, based on the options provided.
The escrow is only enabled for ERC20 fee proxy payment network.

### CODE REVIEW

File: packages/payment-processor/src/payment/encoder-approval.ts [hunk 1]
```diff
@@ -218,6 +218,12 @@ export async function isRequestErc20ApprovalWithoutSwapNeeded(
       }
       break;
     case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_FEE_PROXY_CONTRACT:
+      if (options?.isEscrow) {
+        if (!(await hasEscrowErc20Approval(request, from))) {
+          return true;
+        }
+        return false;
```
Comment (leoslr) [inline packages/payment-processor/src/payment/encoder-approval.ts:225]:
```suggestion
        return !(await hasEscrowErc20Approval(request, from));
```

**Review by leoslr — APPROVED** [review]

File: packages/payment-detection/src/erc20/fee-proxy-contract.ts [hunk 2]
```diff
@@ -19,6 +19,11 @@ const PROXY_CONTRACT_ADDRESS_MAP = {
   ['0.2.0']: '0.2.0',
 };
 
+const ESCROW_CONTRACT_ADDRESS_MAP = {
+  ['0.1.0']: '0.1.0',
+  ['0.2.0']: '0.1.0',
```
Comment (yomarion) [inline packages/payment-detection/src/erc20/fee-proxy-contract.ts:24]: Do we really need to support versions that have not been used by public?

File: packages/payment-processor/src/payment/encoder-payment.ts [hunk 3]
```diff
@@ -41,10 +43,21 @@ export function encodeRequestPaymentWithoutSwap(
         ...overrides,
       };
     case ExtensionTypes.ID.PAYMENT_NETWORK_ERC20_FEE_PROXY_CONTRACT:
-      return {
-        ...prepareErc20FeeProxyPaymentTransaction(request, amount, feeAmount),
-        ...overrides,
-      };
+      if (options?.isEscrow) {
```
Comment (yomarion) [inline packages/payment-processor/src/payment/encoder-payment.ts:46]: I'm not comfortable with that choice. I think escrow is, for now, a minor option, and later probably a separate RN, so I think it's not worth it re-using the same `encodeRequestPayment` method, but instead make a specific one like `encodeRequestPaymentWithEscrow`.

File: packages/payment-processor/src/payment/erc20.ts [hunk 4]
```diff
@@ -159,20 +174,23 @@ export function prepareApproveErc20(
  * the request in its payment currency. Can be used with a Multisig contract.
  * @param request the request to pay
  * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param isEscrow wether to approve the payment proxy or escrow contract
  */
 export function encodeApproveErc20(
   request: ClientTypes.IRequestData,
   signerOrProvider: providers.Provider | Signer = getProvider(),
+  isEscrow = false,
```
Comment (yomarion) [inline packages/payment-processor/src/payment/erc20.ts:182]: Same here, not comfortable with that choice. I think escrow is, for now, a minor option, and later probably a separate RN, so I think it's not worth it re-using the same method.

File: packages/payment-processor/test/payment/encoder-approval.test.ts [hunk 5]
```diff
@@ -493,3 +493,38 @@ describe('Approval encoder handles Eth Requests', () => {
...
+    );
+    const escrowProxyAddress = getEscrowProxyAddress(validRequestERC20FeeProxy);
+    expect(approvalTransaction).toEqual({
+      data: erc20ApprovalData(escrowProxyAddress),
+      to: erc20ContractAddress,
+      value: 0,
+    });
+  });
+
+  it('Should not return anything', async () => {
```
Comment (yomarion) [inline packages/payment-processor/test/payment/encoder-approval.test.ts:513]:
```suggestion
  it('Should not return a transaction if the token is already approved', async () => {
```

**Review by yomarion — COMMENTED** [review]
I'm not sure about the interface override `isEscrow`, can we discuss that?

**Pull request closed without merging**